flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -629,14 +627,9 @@ private void queueChannel(InputChannel channel) {
 
 			if (availableChannels == 0) {
 				inputChannelsWithData.notifyAll();
-				toNotify = isAvailable;
-				isAvailable = AVAILABLE;
+				resetAvailable();
 			}
 		}
-
-		if (toNotify != null) {
-			toNotify.complete(null);
-		}
```

Comment:
I noticed the change here but have not thought through whether it would cause deadlocks before. I would double check it and take another way if truth.

Reply:
I've seen quite a bit deadlocks from exactly this kind of pattern, maybe not in this exact place. I think it's a generally speaking good rule of trying to decouple/release locks from the callbacks. We can not release the higher level locks (from the callers), but still it would make our live a bit easier in the future, if we hold as few locks as possible.

Reply:
Agree, I would get ride of the lock here.